allennlp/data/data_loaders/multi_process_data_loader.py

```diff
@@ -340,24 +358,67 @@ def _instance_worker(self, worker_id: int, queue: mp.JoinableQueue) -> None:
                                 "so already)."
                             )
                     checked_for_token_indexers = True
-                queue.put((instances_chunk, None))
+                queue.put((instance, None))
         except Exception as e:
             queue.put((None, (e, traceback.format_exc())))
 
         # Indicate to the consumer that this worker is finished.
         queue.put((None, None))
 
-        # Wait for consumer to finish to avoid prematurely closing the queue.
+        # Wait until this process can safely exit.
+        queue.join()
+
+    def _batch_worker(self, worker_id: int, queue: mp.JoinableQueue) -> None:
+        try:
+            self.reader._set_worker_info(WorkerInfo(self.num_workers, worker_id))
+            instances = self.reader.read(self.data_path)
+            for batch in self._instances_to_batches(instances):
+                queue.put((batch, None))
+        except Exception as e:
+            queue.put((None, (e, traceback.format_exc())))
+
+        # Indicate to the consumer (main thread) that this worker is finished.
+        queue.put((None, None))
+
+        # Wait until this process can safely exit.
         queue.join()
 
+    def _gather_instances(self, queue: mp.JoinableQueue) -> Iterable[Instance]:
+        done_count: int = 0
+        while done_count < self.num_workers:
+            for instance, worker_error in iter(queue.get, (None, None)):
+                if worker_error is not None:
+                    e, tb = worker_error
+                    sys.stderr.write("".join(tb))
```

Comment:
Why not log it?

Reply:
Logging might be disabled

Reply:
Then your process silently exits with an error code, as expected, since you disabled logging.

Do we do this anywhere else like this?

Reply:
We re-throw the exception below, so don't we get a stack trace anyways if the exception is uncaught?

Reply:
Silently exit? I don't think that's ever expected.

If I'm just running a script that uses AllenNLP, I might not enable logging, and so `logging.error` messages would never show up. I'd still expect to see error messages in `stderr` though.

Reply:
This just reminded me I have to make a fix here. There is no guarantee that exceptions will be pickle-able, and so we can't send an exception from a worker process back to the main process. So instead, we just send the traceback as a string and a string representation of the exception back to the main process. So the best we can is print out the traceback to stderr from the main process, since the traceback that the main process creates will not include the traceback from the worker.

Reply:
I don't know if I did a good job explaining that.

It's definitely not ideal. There may be a better way that I'm not aware of yet.

Reply:
I looked up how I solved this problem in [bettermap](http://github.com/allenai/bettermap), and I actually came to the same conclusion you did: https://github.com/allenai/bettermap/blob/master/bettermap.py#L108

I don't remember why, but if both you and past me think it's a good idea, I think it's probably fine :-)

That code though has never had a problem with pickling exceptions. Maybe because it uses `dill` instead of `pickle`?

Reply:
Well I believe most exceptions are pickled just fine, so it's very likely you just never ran into an exception that was not pickle-able.